### ISSUE TITLE
AuthProviders: OAuth 2.0 Implicit Flows should be upgraded to OAuth 2.1 Authentication Code with PKCE

### ISSUE THREAD

**Issue by jburger** [OctopusDeploy/Issues#7141]
# Prerequisites

- [x] I have searched [open](https://github.com/OctopusDeploy/Issues/issues) and [closed](https://github.com/OctopusDeploy/Issues/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed) issues to make sure it isn't already requested
- [x] I have discussed this ~in #feature-discussion or #backlog and the consensus is that this is something we plan on adding in the near future~ with several key product team members and with Principal Engineers
- [x] I have written a descriptive issue title
- [x] I have linked the original source of this feature request
- [x] I have labelled the value stream (area/core, area/steps, ...)
- [x] I have added the kind/enhancement label

# The enhancement

## What is the problem this solves or benefit it gives

The primary benefit to updating this is that our customers security teams are working to ensure that the vendors they use, meet standards of protocols that are used in application design, like TLS or OAuth. Several customers have indicated that they are focused on deprecated parts of the OAuth 2.0 standard, like the one our application currently uses (Implicit Flow).

In brief technical terms, Implicit Flow makes assumptions about the security of the transport between the browser the STS, and on mobile in particular this is considered a genuine risk. For web applications like ours however, this is not considered an urgent issue, however the standard for implicit flow is still considered deprecated as a result of the new flow that eliminates risk on mobile devices. 

To recap, this will not fix any particular security flaw, but it will give customers confidence that we are keeping up to date with key protocols and standards within our application.

## Proposed solution

There is a POC done here that uplifts the OIDC Authentication provider: https://github.com/OctopusDeploy/OpenIDConnectAuthenticationProviders/pull/52 however it doesn't use PKCE, and further testing is required to make sure it doesn't break Octopus ID. 

Ideally this would be available to all Auth Providers that use our OIDC implementation as a base.

## Internal Links
[One of several issues raised by customers here](https://octopusdeploy.slack.com/archives/CB2QY9CTD/p1622208721008100)

**Comment by N-lson** [comment]
Release Note: Octopus now supports OAuth code flow with PKCE when a client secret is specified

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2022.2 | 2022.2.4501 |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2022.2 | 2022.2.4501 |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2022.2 | 2022.2.4500 |

**Comment by N-lson** [comment]
More internal discussion: https://octopusdeploy.slack.com/archives/C0K9DNQG5/p1621324246008100

**Comment by N-lson** [comment]
FYI - PKCE flow will be enabled if a client secret is specified. If you are moving from implicit flow to PKCE flow, no other changes need to be made and the instance does not need to be restarted.

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2022.2 | 2022.2.4500 |
| 2022.3+ | _all releases_ |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2022.2 | 2022.2.4500 |
| 2022.3+ | _all releases_ |
